substrate/client/network/sync/src/strategy.rs

```diff
@@ -0,0 +1,463 @@
+	pub fn on_blocks_processed(
+		&mut self,
+		imported: usize,
+		count: usize,
+		results: Vec<(Result<BlockImportStatus<NumberFor<B>>, BlockImportError>, B::Hash)>,
+	) {
+		match self {
+			SyncingStrategy::WarpSyncStrategy(_) => {},
+			SyncingStrategy::StateSyncStrategy(strategy) =>
+				strategy.on_blocks_processed(imported, count, results),
+			SyncingStrategy::ChainSyncStrategy(strategy) =>
+				strategy.on_blocks_processed(imported, count, results),
+		}
+	}
+
+	/// Notify a syncing strategy that a block has been finalized.
+	pub fn on_block_finalized(&mut self, hash: &B::Hash, number: NumberFor<B>) {
+		match self {
+			SyncingStrategy::WarpSyncStrategy(_) => {},
+			SyncingStrategy::StateSyncStrategy(_) => {},
+			SyncingStrategy::ChainSyncStrategy(strategy) =>
+				strategy.on_block_finalized(hash, number),
+		}
+	}
+
+	/// Inform sync about a new best imported block.
+	pub fn update_chain_info(&mut self, best_hash: &B::Hash, best_number: NumberFor<B>) {
+		match self {
+			SyncingStrategy::WarpSyncStrategy(_) => {},
+			SyncingStrategy::StateSyncStrategy(_) => {},
+			SyncingStrategy::ChainSyncStrategy(strategy) =>
+				strategy.update_chain_info(best_hash, best_number),
+		}
+	}
+
+	// Are we in major sync mode?
+	pub fn is_major_syncing(&self) -> bool {
+		match self {
+			SyncingStrategy::WarpSyncStrategy(_) => true,
+			SyncingStrategy::StateSyncStrategy(_) => true,
+			SyncingStrategy::ChainSyncStrategy(strategy) =>
+				strategy.status().state.is_major_syncing(),
+		}
+	}
+
+	/// Get the number of peers known to the syncing strategy.
+	pub fn num_peers(&self) -> usize {
+		match self {
+			SyncingStrategy::WarpSyncStrategy(strategy) => strategy.num_peers(),
+			SyncingStrategy::StateSyncStrategy(strategy) => strategy.num_peers(),
+			SyncingStrategy::ChainSyncStrategy(strategy) => strategy.num_peers(),
+		}
+	}
+
+	/// Returns the current sync status.
+	pub fn status(&self) -> SyncStatus<B> {
+		match self {
+			SyncingStrategy::WarpSyncStrategy(strategy) => strategy.status(),
+			SyncingStrategy::StateSyncStrategy(strategy) => strategy.status(),
+			SyncingStrategy::ChainSyncStrategy(strategy) => strategy.status(),
+		}
+	}
+
+	/// Get the total number of downloaded blocks.
+	pub fn num_downloaded_blocks(&self) -> usize {
+		match self {
+			SyncingStrategy::WarpSyncStrategy(_) => 0,
+			SyncingStrategy::StateSyncStrategy(_) => 0,
+			SyncingStrategy::ChainSyncStrategy(strategy) => strategy.num_downloaded_blocks(),
+		}
+	}
+
+	/// Get an estimate of the number of parallel sync requests.
+	pub fn num_sync_requests(&self) -> usize {
+		match self {
+			SyncingStrategy::WarpSyncStrategy(_) => 0,
+			SyncingStrategy::StateSyncStrategy(_) => 0,
+			SyncingStrategy::ChainSyncStrategy(strategy) => strategy.num_sync_requests(),
+		}
+	}
+
+	/// Get actions that should be performed by the owner on the strategy's behalf
+	#[must_use]
+	pub fn actions(&mut self) -> Box<dyn Iterator<Item = SyncingAction<B>>> {
```

Comment:
Yes, we can't return just `impl Iterator` because different strategies return different `impl Iterator`'s, and even though we apply `.map()` to them to convert to the iterator of `StrategyAction`'s, these different mapped iterators remember what was the underlying iterator type and are essentially different types.